### PR TITLE
feat: Added calibration curve with multi-class classification support

### LINKED ISSUE #1316: feat(skore): Display the calibration curve for models
### Is your feature request related to a problem? Please describe.

As a Data Scientist, to inspect my model, I want to see the calibration curve. 

### Describe the solution you'd like

See everything here: https://github.com/probabl-ai/skore/pull/1315  

### Describe alternatives you've considered, if relevant

1. extend to multiclass. It is possible indeed, cf [this example](https://scikit-learn.org/stable/auto_examples/calibration/plot_calibration_multiclass.html#sphx-glr-auto-examples-calibration-plot-calibration-multiclass-py). I personally think it's a bit difficult to read, but maybe it's only because I'm not used to this and it should be added for the people who are working regularly with multiclass.  
2. integrate this also in comparisonReport
  a. I suggest we do it indeed, but in another issue 

### Additional context

sister issue https://github.com/probabl-ai/skore/issues/1317

### PR DESCRIPTION
**Pull request by waridrox**
Closes #1316, Waiting on iteration after merge conflicts that will arise from #1669, #1701, #1709, etc.

Added CalibrationCurveDisplay class, with visualization for classifier probability calibration assessment. Implementation follows scikit-learn's API patterns ([ref](https://scikit-learn.org/stable/auto_examples/calibration/plot_calibration_multiclass.html#sphx-glr-auto-examples-calibration-plot-calibration-multiclass-py)) while extending functionality to support both binary and multiclass classification.

### Summary:

- Implemented CalibrationCurveDisplay class adapting usage from scikit-learn
- Added support for both binary and multiclass classification.
- Implemented different bin strategies ("uniform" and "quantile")
- Introduced DisplayClassProtocol for type safety across display classes (This was introduced to deal with Union types and same class method names).

### Testing
Added tests to cover binary classification scenarios.
Added tests to cover multi-class classification scenarios.

### TODO:
* [ ] Revise documentation and example file (ref - https://gist.github.com/waridrox/eb4fadabf8dd93200c1746161f74e5ca)

___

Example file calibration plots:
1. Single Estimator Calibration Curve
```python
# Calibration curve - Logistic Regression
# Create the figure and axes
fig, (ax, hist_ax) = plt.subplots(
    nrows=2, figsize=(8, 8), height_ratios=[2, 1], sharex=True
)

# Plot the calibration curve
display.plot(
    ax=ax,
    hist_ax=hist_ax,
    line_kwargs={"color": "darkblue", "marker": "o", "linewidth": 2},
    ref_line_kwargs={
        "color": "black",
        "linestyle": "--",
        "alpha": 0.8,
        "linewidth": 1.5,
    },
    hist_kwargs={"color": "skyblue", "alpha": 0.5},
)
```
Screenshot 2025-05-25 at 3 57 49 PM

___

2. Multiple Estimators Comparison
```python
# Calibration curves for different classifiers
# Compares: Logistic Regression, Naive Bayes, Random Forest, and SVC
# Create plot with custom styling
fig, (ax, hist_ax) = plt.subplots(
    nrows=2, figsize=(10, 10), height_ratios=[2, 1], sharex=True
)

# Plot with custom styling
display.plot(
    ax=ax,
    hist_ax=hist_ax,
    line_kwargs={"linewidth": 2, "alpha": 0.8, "marker": "o", "markersize": 5},
    ref_line_kwargs={
        "color": "black",
        "linestyle": "--",
        "alpha": 0.7,
        "linewidth": 1.5,
    },
    hist_kwargs={"alpha": 0.3},
)
```

Screenshot 2025-05-25 at 3 58 05 PM

___

3. Calibrated vs Uncalibrated Models
```python
# Calibration curves for uncalibrated vs. calibrated models
# Compares: Random Forest with different calibration methods (uncalibrated, sigmoid, isotonic)
# Create a nice plot
fig, (ax, hist_ax) = plt.subplots(
    nrows=2, figsize=(10, 10), height_ratios=[2, 1], sharex=True
)

# Plot with custom styling
display.plot(
    ax=ax,
    hist_ax=hist_ax,
    line_kwargs={"linewidth": 2, "alpha": 0.8, "marker": "o", "markersize": 6},
    ref_line_kwargs={
        "color": "black",
        "linestyle": "--",
        "alpha": 0.7,
        "linewidth": 1.5,
    },
    hist_kwargs={"alpha": 0.3},
)
```
Screenshot 2025-05-25 at 3 58 24 PM

___

4. Different Binning Strategies
```python
# Calibration plots for different strategies and bin counts
# Compares: Different binning strategies (uniform vs quantile) with different bin counts
# Setup the figure with GridSpec
fig = plt.figure(figsize=(12, 10))
gs = GridSpec(3, 2, height_ratios=[2, 1, 1])

# Create a single plot for all calibration curves
ax_calibration_curve = fig.add_subplot(gs[0, :])
ax_calibration_curve.set_title(
    "Calibration plots for different strategies and bin counts"
)
```
Screenshot 2025-05-25 at 3 58 34 PM

___

5. Sklearn-Style Comparison (with GridSpec)
```python
# Calibration plots
# Compares: Multiple classifiers in scikit-learn style layout
# Setup the figure with GridSpec
fig = plt.figure(figsize=(12, 10))
gs = GridSpec(4, 2, height_ratios=[2, 2, 1, 1])

# Create a single plot for all calibration curves
ax_calibration_curve = fig.add_subplot(gs[:2, :])
ax_calibration_curve.set_title("Calibration plots", fontsize=16)
```
Screenshot 2025-05-25 at 3 58 46 PM

____

6. Multi-class classification
```python
# Calibration curves for multiclass classification
# Shows: Calibration curves for a three-class classification problem
# Setup the figure with GridSpec
fig = plt.figure(figsize=(15, 10))
gs = GridSpec(3, 3, height_ratios=[2, 1, 1])

# Create a single plot for all calibration curves
ax_calibration_curve = fig.add_subplot(gs[0, :])
ax_calibration_curve.set_title(
    "Calibration curves for multiclass classification", fontsize=16
)
```
Screenshot 2025-05-25 at 3 59 04 PM

___
> Would need to identify what PRs currently in-progress might affect this before iterating on something critical. Hence, currently setting this to a draft version.
> CC: @MarieSacksick @auguste-probabl

### CODE REVIEW

**Comment by glemaitre** [comment]
In terms of API, I think that we will want to have the calibration display in another accessor than metric. It is more of a diagnosis tool so we need to think where to add it.

I'm also thinking that we should enlarge the scope of the tool potentially and call it a reliability diagram that would cover the classification and regression case.

Also I did not yet go into the code but we need to think twice before implementing the multiclass case. Use a one vs rest approach might not be ideal since some output of classifiers are not using this strategy as a prediction function. I was looking at https://arxiv.org/pdf/2112.10327 and it seems that you have several trade-off and we should look at it thoroughly.

**Comment by thomass-dev** [comment]
Please update your PR with main, so that the `pytest` workflow status will be reported.

**Comment by thomass-dev** [comment]
Please update your PR with main, so that the `pytest` workflow status will be reported.

**Comment by glemaitre** [comment]
The design moved quite a lot since and we have now the first calibration plot for the case of an estimator. I'm closing since it would be easier to restart from `main`.
